aws/ci_builder.rb

```diff
@@ -0,0 +1,80 @@
+require 'cdo/rake_utils'
+require_relative 'ci_build_metrics'
+module CiBuilder
+  require_relative '../lib/cdo/rake_utils'
+  extend self
+  BUILD_STARTED = 'build-started'.freeze
+
+  def build
+    Dir.chdir(deploy_dir) do
+      return 0 unless build_required?
+
+      FileUtils.touch BUILD_STARTED
+
+      git_update_count = update_repository
+
+      log_build_history(git_update_count)
+
+      display_commit_link
+
+      install_dependencies
+
+      notify_honeybadger
+
+      execute_ci_task
+    end
+  end
+
+  private
+
+  def build_required?
+    RakeUtils.git_updates_available? || File.file?(BUILD_STARTED) || !CDO.daemon
+  end
+
+  def update_repository
+    RakeUtils.git_fetch
+    count = RakeUtils.git_update_count
+    RakeUtils.git_pull if count > 0
+    [1, count].max # I think this is not needed. unless count can be a real number between 0 and 1.
+  end
+
+  def log_build_history(git_update_count)
+    log = `git log --pretty=format:"%h %s (%an)" -n #{git_update_count}`
+    File.write(deploy_dir('rebuild'), log)
+  end
```

Comment:
How about merging these two methods into one? I kinda like the idea of having just one method that makes changes to the filesystem as a side effect of being called, and I *really* like the idea of not needing to persist the ephemeral data point of "how many changes are we about to pull" between multiple methods. Instead, we can just capture it all in one place:

```suggestion
  def update_repository
    RakeUtils.git_fetch
    update_count = RakeUtils.git_update_count
    RakeUtils.git_pull if update_count > 0
    update_log = `git log --pretty=format:"%h %s (%an)" -n #{[1, update_count].max}`
    File.write(deploy_dir('rebuild'), )
  end
```